File: BlackjackApp/Domain/Models/SideBet.swift
import Foundation

// MARK: – Perfect Pairs

enum PerfectPairResult: String, Sendable {
    case none
    case mixedPair    // same rank, different suit colours       — payout 5:1
    case colouredPair // same rank, same colour, different suits — payout 10:1
    case perfectPair  // same rank, identical suit               — payout 25:1
}

// MARK: – 21+3 (player's 2 cards + dealer upcard)

enum TwentyOnePlusThreeResult: String, Sendable {
    case none
    case flush         // same suit, no sequence              — payout 5:1
    case straight      // sequential ranks, mixed suits       — payout 10:1
    case threeOfAKind  // same rank, different suits          — payout 30:1
    case straightFlash // sequential ranks, same suit         — payout 40:1
    case suitedTrips   // same rank AND same suit             — payout 100:1
}

// MARK: – Generic containers

enum SideBetType: String, CaseIterable, Sendable {
    case perfectPairs
    case twentyOnePlusThree
}

struct SideBetWager: Identifiable, Sendable {
    let id: UUID        = UUID()
    let type: SideBetType
    let amount: Double
}

struct SideBetResult: Identifiable, Sendable {
    let id: UUID          = UUID()
    let wager: SideBetWager
    let perfectPair:        PerfectPairResult?
    let twentyOnePlusThree: TwentyOnePlusThreeResult?

    /// Net payout including return of wager (positive = win, 0 = push, negative = loss).
    var netPayout: Double {
        let stake = wager.amount
        switch wager.type {
        case .perfectPairs:
            switch perfectPair {
            case .mixedPair:    return stake * 5
            case .colouredPair: return stake * 10
            case .perfectPair:  return stake * 25
            default:            return -stake
            }
        case .twentyOnePlusThree:
            switch twentyOnePlusThree {
            case .flush:         return stake * 5
            case .straight:      return stake * 10
            case .threeOfAKind:  return stake * 30
            case .straightFlash: return stake * 40
            case .suitedTrips:   return stake * 100
            default:             return -stake
            }
        }
    }
}

File: src/domain/models/sideBet.ts
export enum PerfectPairResult {
  NONE          = 'none',
  MIXED_PAIR    = 'mixedPair',    // same rank, different colours  — payout 5:1
  COLOURED_PAIR = 'colouredPair', // same rank, same colour        — payout 10:1
  PERFECT_PAIR  = 'perfectPair',  // same rank, identical suit     — payout 25:1
}

export enum TwentyOnePlusThreeResult {
  NONE           = 'none',
  FLUSH          = 'flush',         // same suit, no sequence        — payout 5:1
  STRAIGHT       = 'straight',      // sequential ranks, mixed suits — payout 10:1
  THREE_OF_KIND  = 'threeOfAKind',  // same rank, different suits   — payout 30:1
  STRAIGHT_FLUSH = 'straightFlush', // sequential ranks, same suit  — payout 40:1
  SUITED_TRIPS   = 'suitedTrips',   // same rank AND same suit      — payout 100:1
}

export enum SideBetType {
  PERFECT_PAIRS         = 'perfectPairs',
  TWENTY_ONE_PLUS_THREE = 'twentyOnePlusThree',
}

export interface SideBetWager {
  readonly id: string
  readonly type: SideBetType
  readonly amount: number
}

export interface SideBetResult {
  readonly id: string
  readonly wager: SideBetWager
  // Invariant: exactly one is non-null, matching wager.type
  readonly perfectPair:        PerfectPairResult | null
  readonly twentyOnePlusThree: TwentyOnePlusThreeResult | null
}

// Net payout (positive = win, negative = loss). Does NOT include stake return.
export function sideBetNetPayout(result: SideBetResult): number {
  const stake = result.wager.amount
  if (result.wager.type === SideBetType.PERFECT_PAIRS) {
    switch (result.perfectPair ?? PerfectPairResult.NONE) {
      case PerfectPairResult.MIXED_PAIR:    return stake * 5
      case PerfectPairResult.COLOURED_PAIR: return stake * 10
      case PerfectPairResult.PERFECT_PAIR:  return stake * 25
      default:                              return -stake
    }
  } else {
    switch (result.twentyOnePlusThree ?? TwentyOnePlusThreeResult.NONE) {
      case TwentyOnePlusThreeResult.FLUSH:          return stake * 5
      case TwentyOnePlusThreeResult.STRAIGHT:       return stake * 10
      case TwentyOnePlusThreeResult.THREE_OF_KIND:  return stake * 30
      case TwentyOnePlusThreeResult.STRAIGHT_FLUSH: return stake * 40
      case TwentyOnePlusThreeResult.SUITED_TRIPS:   return stake * 100
      default:                                       return -stake
    }
  }
}
